src/runtime_src/core/edge/user/aie/graph_object.cpp

```diff
@@ -0,0 +1,303 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2024 Advanced Micro Devices, Inc. All rights reserved.
+
+#include "graph_object.h"
+#include "core/common/system.h"
+#include <memory>
+#include "core/edge/user/shim.h"
+
+namespace zynqaie {
+  graph_object::graph_object(ZYNQ::shim* shim, const xrt::uuid& uuid , const char* gname,
+                  xrt::graph::access_mode am, const zynqaie::hwctx_object* hwctx)
+    : m_shim{shim},
+      name{gname},
+      access_mode{am}
+  {
+      auto device{xrt_core::get_userpf_device(m_shim)};
+      auto drv = ZYNQ::shim::handleCheck(device->get_device_handle());
+
+      if (!drv->isAieRegistered())
+        throw xrt_core::error(-EINVAL, "No AIE presented");
+      aieArray = drv->getAieArray();
+      id = xrt_core::edge::aie::get_graph_id(device.get(), name, hwctx);
+      if (id == xrt_core::edge::aie::NON_EXIST_ID)
+        throw xrt_core::error(-EINVAL, "Can not get id for Graph '" + name + "'");
+
+      int ret = drv->openGraphContext(uuid.get(), id, am);
+      if (ret)
+        throw xrt_core::error(ret, "Can not open Graph context");
+
+      /* Initialize graph tile metadata */
+      graph_config = xrt_core::edge::aie::get_graph(device.get(), name, hwctx);
+
+      /* Initialize graph rtp metadata */
+      rtps = xrt_core::edge::aie::get_rtp(device.get(), graph_config.id, hwctx);
+      pAIEConfigAPI = std::make_shared<adf::graph_api>(&graph_config);
+      pAIEConfigAPI->configure();
+      state = graph_state::reset;
+      drv->getAied()->register_graph(this);
+  }
+
+  graph_object::~graph_object()
+  {
+    auto device{xrt_core::get_userpf_device(m_shim)};
```

Comment:
I wish we could get rid of class shim wrapping xrt_core::device.      This is a much bigger project though, but now that we no longer support shim APIs and linking with shim, it's doable and much cleaner.   I think a good example is XRT-MCDM where shim and device class derived from xrt_core::device is one and the same.

Reply:
This could be bigger change , we can handle in separate PR.